annotate.el

```diff
@@ -1073,6 +1090,116 @@ example:
+ "Update datbase *on disk* removing all the records with empty
+annotation."
+  (interactive)
+  (let ((db (annotate-db-clean-records (annotate-load-annotation-data))))
+    (annotate-dump-annotation-data db)))
+
+(defun annotate-load-annotation-data ()
+  "Read and return saved annotations."
+  (with-temp-buffer
+    (when (file-exists-p annotate-file)
+      (insert-file-contents annotate-file))
+    (goto-char (point-max))
+    (cond ((= (point) 1)
+           nil)
+          (t
+           (goto-char (point-min))
+           (read (current-buffer))))))
+
+(defun annotate-dump-annotation-data (data)
+  "Save `data` into annotation file."
+  (with-temp-file annotate-file
+    (let ((print-length nil))
+      (prin1 data (current-buffer)))))
+
+(defun annotate-db-remove-annotation (db-records
+                                      record-filename
+                                      annotation-beginning
+                                      annotation-ending)
+  "Remove from database `db-records' the annotation indentified by
+ the triplets `record-filename', `annotation-beginning' and
+ `annotation-ending'; if such annotation does exists."
+  (let* ((filename-match-p (lambda (record)
+                             (string= (annotate-filename-from-dump record)
+                                      record-filename)))
+         (file-matched-record (cl-find-if filename-match-p db-records)))
+    (if file-matched-record
+        (let* ((rest-of-db      (cl-remove-if filename-match-p db-records))
+               (new-annotations (cl-remove-if (lambda (a)
+                                                (and (= (annotate-beginning-of-annotation a)
+                                                        annotation-beginning)
+                                                     (= (annotate-ending-of-annotation    a)
+                                                        annotation-ending)))
+                                              (annotate-annotations-from-dump file-matched-record)))
+               (checksum        (annotate-checksum-from-dump file-matched-record))
+               (new-record      (annotate-make-record record-filename
+                                                      new-annotations
+                                                      checksum)))
+          (push new-record
+                rest-of-db))
+      db-records)))
+
+(defun annotate-db-replace-annotation (db-records
+                                       record-filename
+                                       annotation-beginning
+                                       annotation-ending
+                                       replacing-text)
+  "TODO"
+  (let* ((filename-match-p (lambda (record)
+                             (string= (annotate-filename-from-dump record)
+                                      record-filename)))
+         (file-matched-record (cl-find-if filename-match-p db-records)))
+    (if file-matched-record
```

Comment:
In other programming languages, I would insert a `if not file-matched-record: return db-records` here, in order to reduce the nesting.

Perhaps this could be done in lisp as well by inlining the `if` into the `let`, i.e. `(let ((old-annotation (if file-matched-record (cl-find-if ...) nil))))`. Or even `(let ((old-annotation (and file-matched-record (cl-find-if ...)))))`.

Again, I don't know if this is more or less idiomatic, but at least it reduces some of the nesting, and possibly the trailing default return values.

Reply:
Usually (but is not a rule cast in stone :-)) i prefer `(if something ...` to `(if (not something...`, just less cognitive burden for me (if i understood correctly what you wrote); if you prefer the form the `not` i could change that as the test is very simple (the simplest possible :) ).

I kind to avoid  complex forms when binding variable: but, again this are just stylistic preferences.

Bye!
C.

Reply:
Interesting!

I think I was advocating that it might be better to push complexity into the variable binding to avoid complexity in the ensuing code. You are saying exactly the opposite.

That's fascinating. I see merits in both ways of looking at it.

I wish we could meet for a beer one day and just talk about programming for a while ;-).